### ISSUE TITLE
[2.5] Audio delay on fresh install when running three parallel Kurento media servers

### ISSUE THREAD

**Issue by mokazemi** [bigbluebutton/bigbluebutton#15172]
I was installing BBB 2.5 on a new server and dealing with high audio delay (**audio and video where totally out of sync**).
I had copied `/etc/bigbluebutton/bbb-conf/apply-config.sh` file from the old server to the new one, without much changing it. There was called the function `enableMultipleKurentos` in it, [based on here](https://docs.bigbluebutton.org/admin/customize.html#run-three-parallel-kurento-media-servers) to run three parallel KMS. 

I reinstalled it again and noticed that using `enableMultipleKurentos` function in `apply-config.sh` file may be causing this problem. 
adding `disableMultipleKurentos` didn't solve the problem. So I ended up removing all files and installing again, without adding that function into apply-config fie. And it's solved now. 

P.S. I'm not sure if it's been cause of the problem! I just decided to report it here, for more investigations. 
Since I'm not currently able to reproduce it on another server.
 
**To Reproduce**
1. fresh installation of BBB 2.5
2. add `enableMultipleKurentos`  to file `/etc/bigbluebutton/bbb-conf/apply-config.sh`
3. Run `sudo bbb-conf --restart`
4. Turn on audio and camera
5. Check on another device to see if it has any delay or not

Version: **BigBlueButton Server 2.5.1 (3020)**

**Comment by defnull** [comment]
This may be caused by https://github.com/bigbluebutton/bigbluebutton/pull/14893

**Comment by mokazemi** [comment]
> This may be caused by #14893

I'm not sure, but I'd used `data="jitterbuffer_msec=60:120:20"` [based on this](https://groups.google.com/g/bigbluebutton-setup/c/3Y7VBllwpX0/m/e05YaXQWBgAJ).

**Comment by defnull** [comment]
Other than documented, looking at the actual implementation, the third parameter seems to be ignored: https://github.com/signalwire/freeswitch/blob/master/src/mod/endpoints/mod_sofia/rtp.c#L626 -> maxlen is set if there is a colon, but that's it. A second colon does not do anything.

**Comment by defnull** [comment]
Also, `<action application="jitterbuffer" data="60:120" />` and `<action application="set" data="jitterbuffer_msec=60:120" />` do different things if I understand that correctly. The first one controls the RTP jitterbuffer within freeswitch, the second one sets parameters for the opus codec, which also has a jitterbuffer. They may both have an impact.

**Comment by mokazemi** [comment]
Thanks for mentioning this. I hadn't noticed the difference. 
By the way, I wonder why the problem is solved after not calling `enableMultipleKurentos` function. 
I thought it might do some overwrite that causes media to become out of sync, since I've read that BBB 2.5 uses mediasoup instead. (I'm not a developer, so it's just a guess!)

**Comment by defnull** [comment]
We changed the setting back to `<action application="jitterbuffer" data="60:120" />` and could not reproduce the high-delay issue since then. Still not 100%, but we'll keep the `60:120` setting for now.

**Comment by mokazemi** [comment]
> We changed the setting back to `<action application="jitterbuffer" data="60:120" />` and could not reproduce the high-delay issue since then. Still not 100% sure, but we'll keep the `60:120` setting for now.


This jitterbuffer setting doesn't cause any problem for me. Since it's now 120 and everything is alright.  
and when I was facing described problem, it was 60:120, and still not fixed (more than 2 seconds delay.)

**Comment by ho30hero** [comment]
Hello
We also have a relatively similar problem but without any change or modification to settings and configuration files. 
After installing new fresh BBB 2.5.4 on a dedicated server (12 core + 64 ram)  (also I tested before in vps ( 8 core + 16 ram) with the same issue):
- Sometimes when we have about 3-5 concurrent web came sharing with presentation enabled, we have unsynced sound for about 3-10 seconds delay from the video, and it may be continued for a while for some of the attendees or moderators from 1 minute to much more minutes. 
- No special Errors in --status and --check but this one : 
   [https://groups.google.com/g/bigbluebutton-setup/c/T2bDkdZXLT8](https://groups.google.com/g/bigbluebutton-setup/c/T2bDkdZXLT8)
Regards

**Comment by MobinDev** [comment]
> Hello We also have a relatively similar problem but without any change or modification to settings and configuration files. After installing new fresh BBB 2.5.4 on a dedicated server (12 core + 64 ram) (also I tested before in vps ( 8 core + 16 ram) with the same issue):
> 
> * Sometimes when we have about 3-5 concurrent web came sharing with presentation enabled, we have unsynced sound for about 3-10 seconds delay from the video, and it may be continued for a while for some of the attendees or moderators from 1 minute to much more minutes.
> * No special Errors in --status and --check but this one :
>   https://groups.google.com/g/bigbluebutton-setup/c/T2bDkdZXLT8
>   Regards

I have exactly this problem on dedicated server (32 core CPU  + 16 GB Ram)

**Comment by ffdixon** [comment]
Can you check the audio delay on 

  https://test25.bigbluebutton.org/

and let us know if you encounter the same delay.  You can use

 https://test25.bigbluebutton.org/demo/demo3.jsp

to join with multiple users.

**Comment by MobinDev** [comment]
> Can you check the audio delay on
> 
> https://test25.bigbluebutton.org/
> 
> and let us know if you encounter the same delay. You can use
> 
> https://test25.bigbluebutton.org/demo/demo3.jsp
> 
> to join with multiple users.

This does not always happen to everyone!  But I have a session recorded on my server that shows it happening to one person in a session with 6 people.

**Comment by sonerd** [comment]
we also have the delay Problem since 2.5.5. Sometimes it happens in meetings with only two users and some times with more.

**Comment by lordwebbie** [comment]
I am also following this issue, because we have these problems in some conferences. We were able to find out that people suffering from high delay, at some time in the conference have had very short connection issues (e.g. high ping or high packet loss). This often was due to an unstable wifi connection. It made us curious, why the problems did not happen with 2.4, but with 2.5.x only.

What was the exact intention of https://github.com/bigbluebutton/bigbluebutton/pull/14893/commits/145307d4dbf62e557a9b65e911d4e94adb69e19e?

The documentation (https://freeswitch.org/confluence/display/FREESWITCH/JitterBuffer) states: "The jitter buffer has three params that control its behavior: length, max length, and max drift. Length is the initial size of the jitter buffer in milliseconds. Max length is the upper bound for how big the jitter buffer can grow. [...]"

In the referenced commit, @ffdixon changed the JitterBuffer params from "60:120" (Initial: 60ms, Max: 120ms) to "120" (Initial: 120). Then max is not specified and can grow as big as it wants, or am I wrong? So my understanding is, that when there is no upper bound for the jitter buffer, everything works as expected. But when a client has a temporary problem with its network connection (e.g. wifi unstable), then the jitterbuffer might be increased so much, that there is a huuuge delay.

Thus, I do not understand, why the changes to the jitterbuffer have been made....

**Comment by mokazemi** [comment]
Hello. @ffdixon Today I did some experiments on what discussed [here](https://github.com/bigbluebutton/bigbluebutton/issues/15644#issuecomment-1251430382). 

Here's my investigations:

**1. jitter buffer on `120` vs `60:120:20`**

   a) When it's set to `60:120:20` audio is a little bit cracky in case of packet loss, but there's much unusual delay.

   b) When it's set to `120` everything is normal for good connection. **But when there's packet loss**, There would be much **delay** in sending audio. Such when you make a sound, and after two or three seconds the name is indicated at the top of the screen and it receives to the other side. But the video receives without any delay. So it would be out of sync. consequently, people start interrupting each other. 

* NOTE: I wasn't able to simulate packet loss. But it sounds the delay gradually appears in bad connections. 
_________
**2. Server clock out of sync**
- First disable ntp with `timedatectl set-ntp no`
- Then change the time a few seconds with `sudo timedatectl set-time`
- Record something!
- Don't forget to enable ntp after experiment with `timedatectl set-ntp yes`

**Result:**
**There was nothing out of sync in the recording.** I'm not sure what was the problem before. Maybe the problem described [here](https://github.com/bigbluebutton/bigbluebutton/issues/8815) has been something else. I'll try to make more investigations if it happened again.

**Comment by lordwebbie** [comment]
This is exactly what I wrote above. And it all makes sense regarding the issue.

Jitterbuffer will always add a delay, because it is a buffer that can be used when there are very short connection issues.
So you generally want the jitterbuffer to be as small as possible, but as long as it needs to be to overcome connectivity issues. Now, jitterbuffers may be dynamically increased by the application, so people with good connections have low latency, and people with bad connection have higher latency but uninterrupted audio/video.

Problem is: There has to be an upper bound for the jitterbuffer size. It does not make much sense to have a jitterbuffer (= delay) of 4 seconds in real-time communication. Instead, one would accept crackling audio sometimes (especially, if the connectivity issues are only very short, because of wifi or whatever)

So this is why I asked above, why @ffdixon changed the setting with "improving audio" as a commit message. It does not make any sense in my understanding.

**Comment by mokazemi** [comment]
I also noticed the delay gets shorter gradually when the connections become stable. 
So I think that's somehow a correct behavior. But in some other applications (Like Google Meet) the audio plays slower in bad connection (instead of becoming cracky) and the delay disappears right after the connection becomes stable by playing audio a bit faster.

**Comment by invokablegmbh** [comment]
I opened a pull request to set a maximum jitterbuffer again. We are using "100:250" (100ms default, 250ms max) and it works well. These params are increased compared to the original ones @ffdixon changed (60:120).

Basically, if a client has a bad internet connection, the client will gradually increase the jitterbuffer value to be prepared for future connection problems. Jitterbuffers will always add delay.
This issue here is about people suffering from very sporadic connection problems: Not specifiying a maximum results in huge delays, even if a client has just a very short connection issue.

**Comment by sonerd** [comment]
I can confirm that the problem occurs for users with packet loss. During a test meeting where it happened we checked the connection status and saw packet loss at the affected user.

**Comment by invokablegmbh** [comment]
We can confirm, that issues appeared directly after switching from 2.4 to 2.5.4. So we did a check on what changed and came across your jitterbuffer change. We reverted to 60:120 and the problem did not occur anymore. Thus, we had a look at the exact documentation and thought, that it might be a good idea to have the jitterbuffer increased to 100:250, so we are accepting a quarter of a second delay on bad connections. But we do not accept it to grow any further. In worst case this means that there will be crackling audio or some short interruptions. But it turned out that for the majority of the users the connection problems are very very short. So after the connection is OK again, there will be no big delay. This is more satisfying for the users.

Edit: I at first wrote with my personal account. This is why I deleted my first message here.
